File: frontend/src/types/Player.ts
export interface DRTypeAggregateBySec {
  agg_damage: number;
  agg_pre_damage: number;
  type: number;
  citadel_type: number;
  entindex_inflictor: number;
  entindex_ability: number;
  agg_damage_absorbed: number;
  victim_health_max: number;
  victim_health_new: number;
  flags: number;
  ability_id: number;
  attacker_class: number;
  victim_class: number;
  victim_shield_max: number;
  agg_victim_shield_new: number;
  agg_hits: number;
  agg_health_lost: number;
}

export interface DamageRecord {
  damage?: number;
  pre_damage?: number;
  type?: number;
  citadel_type?: number;
  entindex_inflictor?: number;
  entindex_ability?: number;
  damage_absorbed?: number;
  victim_health_max?: number;
  victim_health_new?: number;
  flags?: number;
  ability_id?: number;
  attacker_class?: number;
  victim_class?: number;
  victim_shield_max?: number;
  victim_shield_new?: number;
  hits?: number;
  health_lost?: number;
}

// Comes from DeadlockAPI
// export interface PlayerPathState {
//   player_slot: number;
//   x_min: number;
//   y_min: number;
//   x_max: number;
//   y_max: number;
//   x_pos: number[];
//   y_pos: number[];
//   health: number[];
//   move_type: number[];
//   combat_type: number[];
// }

// victim_id -> DamageRecord[]
export type ParsedVictimDamage = Record<string, DamageRecord[]>;
/**
 * Damage timeline:
 * Index = tick (second). Each element is either:
 *  - ParsedVictimDamage (victim -> damage records)
 *  - null (no damage that tick)
 */
export type Damage = ParsedVictimDamage[];

// Comes from parser
export interface PlayerPosition {
  custom_id: string;
  x: number;
  y: number;
  z: number;
  is_npc: boolean;
}

// All positions for one tick (can contain nulls if a player absent that tick)
export type PositionWindow = PlayerPosition[];

// Comes from DeadlockAPI
export interface PlayerInfo {
  account_id: number;
  player_slot: number;
  team: number;
  hero_id: number;
}

// Comes from parser
export interface PlayerData {
  custom_id: string;
  entity_id: string;
  hero_id: number; // FIXME: ugly but needed for hero lookup
  lane: number;
  lobby_player_slot: number;
  name: string;
  steam_id_32: number;
  team: number;
  zipline_lane_color: number;
  hero: Hero; // FIXME: temporary, to hold hero data
  // player_info: PlayerInfo;
  // path_state: PlayerPathState;
}

// Per-player aggregated data (damage + positions windows)
export interface PlayerGameData {
  positions: PositionWindow;
  damage: Damage;
}

// Temporary ParsedPlayer type (extend if backend adds extra fields)
export interface ParsedPlayer extends PlayerData {
  // Add any additional parsed-only fields here if needed
}

export interface Hero {
  id: number;
  name: string;
  images: { [key: string]: string };
}

// Comes from parser
export interface NPC {
  entity_id: number;
  name: string;
}

File: backend/app/domain/player.py
from typing import Any
from sqlmodel import SQLModel

# PlayerPathState is pulled from Deadlock API (match_metadata call)
# player_slot is the slot in the match, starting from 0.
class PlayerPathState(SQLModel):
    player_slot: int
    x_min: float
    y_min: float
    x_max: float
    y_max: float
    x_pos: list[int]
    y_pos: list[int]
    health: list[float]
    move_type: list[int]
    combat_type: list[int]

# PlayerInfo is pulled from Deadlock API (match_metadata call) and account_id
# is a 32-bit steam_id. The regular/web steam_id is 64-bit.
# player_slot is the slot in the match, starting from 0.
class PlayerInfo(SQLModel):
    account_id: int
    player_slot: int
    team: int
    hero_id: int

class ParsedPlayer(SQLModel):
    entity_id: int
    name: str
    steam_id_32: int

class NPC(SQLModel):
    entity_id: int
    name: str

class Player(SQLModel):
    entity_id: int
    name: str
    steam_id_32: int
    player_info: PlayerInfo
    path_state: PlayerPathState
